### PR TITLE
some cleanup + bug fixes

### PR DESCRIPTION
**Pull request by dengemann**
@jona-sassenhagen 1 pass of cleaning up your PR code. I'm btw. not really happy with the corrmap return signature. The plot is done inside, no figure is returned, etc. We need to think about a better API for that. Maybe we can have a function that takes a list of ICA solutions and plots one figure per label. To be discussed.

### CODE REVIEW

**Comment by jona-sassenhagen** [comment]
Thanks Denis, going over it right now so I understand what you did ... did you check get_ica_map again? You had some doubts earlier.

**Comment by dengemann** [comment]
> did you check get_ica_map again?

Nope ...

File: mne/preprocessing/ica.py [hunk 1]
```diff
@@ -184,6 +184,10 @@ class ICA(ContainsMixin):
         The measurement info copied from the object fitted.
     `n_samples_` : int
         the number of samples used on fit.
+    `labels_` : dict
+        A dictionary of independent component indices, grouped by kinds
+        of artifacts. This attribute is set by some of the artifact detection
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: Mustn't necessarily be artifacts - in theory, you can also identify e.g. central-midline ICs this way, although AFAIK at least CORRMAP has not been used for that purpose so far. Maybe ' A dictionary of independent component indices, grouped by IC type. This attribute is set by some of the artifact detection'.

File: examples/preprocessing/plot_corrmap_detection.py [hunk 2]
```diff
@@ -40,30 +40,32 @@
...
+all_epochs = mne.Epochs(raw, events, event_id, tmin, tmax,
+                        proj=False, picks=picks, baseline=(None, 0),
+                        preload=True, reject=None, verbose=False)
+
+all_epochs = [all_epochs[start:stop] for start, stop in
+              [(0, 100), (101, 200), (201, 300)]]
 
 icas = [ICA(n_components=20, random_state=0).fit(epochs)
-        for epochs in subsets]
+        for epochs in all_epochs]
```
Comment (dengemann) [inline examples/preprocessing/plot_corrmap_detection.py:57]: a bit more explicit ... making parsing easier.

File: examples/preprocessing/plot_corrmap_detection.py [hunk 3]
```diff
@@ -40,30 +40,32 @@
...
 
 # 3) Zeroing the identified blink components for all data sets
-#    results in individually cleaned data sets
+#    results in individually cleaned data sets. Specific components
+#    can be accessed using the label_ attribute.
 
-cleaned_epochs = [ica.apply(epoch, exclude=ica.labels["blinks"], copy=True)
-                  for i, (epoch, ica) in enumerate(zip(subsets, icas))]
+for ica in icas:
+    print(ica.labels_)
```
Comment (dengemann) [inline examples/preprocessing/plot_corrmap_detection.py:71]: somple demo where to find stuff, cleaning is demonstrated else where

File: mne/preprocessing/ica.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 
 from copy import deepcopy
 from inspect import getargspec, isfunction
-from collections import namedtuple, defaultdict
```
Comment (dengemann) [inline mne/preprocessing/ica.py:9]: ... many people don't know it. Better rely on regular python dicts here for the API

File: mne/preprocessing/ica.py [hunk 5]
```diff
@@ -1009,7 +1016,10 @@ def find_bads_eog(self, inst, ch_name=None, threshold=3.0,
         if len(scores) == 1:
             scores = scores[0]
 
-        return eog_idx, scores
+        if not hasattr(self, 'labels_'):
+            self.labels_ = {}
+        self.labels_['eog'] = list(eog_idx)
+        return self.labels_['eog'], scores
```
Comment (dengemann) [inline mne/preprocessing/ica.py:1022]: while we are at it.

File: mne/preprocessing/ica.py [hunk 6]
```diff
@@ -741,7 +745,7 @@ def score_sources(self, inst, target=None, score_func='pearsonr',
             Callable taking as arguments either two input arrays
             (e.g. Pearson correlation) or one input
             array (e. g. skewness) and returns a float. For convenience the
-            most common score_funcs are available via string labels: Currently,
+            most common score_funcs are available via string labels_: Currently,
```
Comment (dengemann) [inline mne/preprocessing/ica.py]: it's an attribtue we set later. Following the sklearn convention for this we let it terminate with an underscore.

File: mne/preprocessing/ica.py [hunk 7]
```diff
@@ -2095,7 +2105,7 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
 
 @verbose
 def corrmap(icas, template, threshold="auto", label="bads",
-            plot=True, inplace=False, ch_type="eeg"):
+            plot=True, copy=False, ch_type="eeg"):
```
Comment (dengemann) [inline mne/preprocessing/ica.py:2108]: we call it copy elsewhere

Comment (dengemann) [inline mne/preprocessing/ica.py:2108]: everywhere else ...

File: mne/preprocessing/ica.py [hunk 8]
```diff
@@ -2154,7 +2164,7 @@ def corrmap(icas, template, threshold="auto", label="bads",
 
     def get_ica_map(ica, components=None):
         if components is None:
-            components = range(ica.n_components_)
+            components = list(range(ica.n_components_))
```
Comment (dengemann) [inline mne/preprocessing/ica.py:2167]: would have bugged in python 3

File: mne/preprocessing/ica.py [hunk 9]
```diff
@@ -2244,7 +2254,7 @@ def _plot_corrmap(data, subjs, indices, ch_type, ica, label):
         plt.show()
 
     if threshold == "auto":
-        threshold = np.arange(60, 95) / 100
+        threshold = np.arange(60, 95, dtype=np.float64) / 100.
```
Comment (dengemann) [inline mne/preprocessing/ica.py:2257]: ... generated zeros ...

File: mne/preprocessing/ica.py [hunk 10]
```diff
@@ -2261,10 +2271,8 @@ def _plot_corrmap(data, subjs, indices, ch_type, ica, label):
         try:
             nt, mt, s, mx = find_max_corrs(all_maps, target, threshold)
         except ValueError:
-            if threshold > 1:
-                logger.info("No component detected using find_outliers."
-                            " Consider using threshold='auto'")
-            return
+            raise RuntimeError("No component detected using find_outliers."
+                               " Consider using threshold='auto'")
```
Comment (dengemann) [inline mne/preprocessing/ica.py]: prefer crashes.

Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: It's not really an error - it's just what happens if the algorithm doesn't pick up on anything; sometimes that's the correct response. It's like ICA not converging, which also raises warnings, not errors.

File: examples/preprocessing/plot_corrmap_detection.py (hunk 3 above)
Comment (jona-sassenhagen) [inline examples/preprocessing/plot_corrmap_detection.py:71]:
```
for ica in icas:
    print(ica.labels_["blinks"])
```

?

**Comment by dengemann** [comment]
@jona-sassenhagen addressed your comments. How would you feel about a separate plot corrmap function? It would take a list of ICAs as input.

**Comment by jona-sassenhagen** [comment]
I'm not sure I get it ... the CORRMAP workflow is usually that you first do one or more "dry" runs (where the real ICA objs. are not modified), where the plot is inspected to see if the parameters (template, threshold etc.) were appropriate. Then, CORRMAP is run again, this time actually committing the labelled ICs; now, the plot is optional. This is done once per artifact type. So in this case, you'd do

```
corrmap(..., inplace=False, plot=True)
corrmap(..., inplace=True, plot=whatever)
```

Couldn't we add the figs to the return values? Or even skip returning the ICs, as it's nondestructive and done in-place, and only return the figure?

**Comment by dengemann** [comment]
@jona-sassenhagen can you merge and we move on in a second iteration?

**Pull request merged**